vmm/src/vstate.rs

```diff
@@ -226,9 +341,159 @@ impl Vcpu {
+        dir.set_file_name("kvm_msrs.json");
+        let reader = BufReader::new(File::open(dir.as_path())?);
+        let entries: Vec<kvm_msr_entry> = serde_json::from_reader(reader)?;
+
+        let vec_size_bytes =
+            std::mem::size_of::<kvm_msrs>() + (entries.len() * std::mem::size_of::<kvm_msr_entry>());
+        let vec: Vec<u8> = Vec::with_capacity(vec_size_bytes);
+        #[allow(clippy::cast_ptr_alignment)]
+        let msrs: &mut kvm_msrs = unsafe {
+            &mut *(vec.as_ptr() as *mut kvm_msrs)
+        };
+
+        unsafe {
+            msrs.entries.as_mut_slice(entries.len()).copy_from_slice(&entries);
+        }
+        msrs.nmsrs = entries.len() as u32;
+        self.fd.set_msrs(msrs)
+    }
+
+    fn setup_regs_from_file(&self, dir: &mut PathBuf) -> result::Result<(), io::Error> {
+        dir.push("kvm_regs.json");
+        let reader = BufReader::new(File::open(dir.as_path())?);
+        let regs: kvm_regs = serde_json::from_reader(reader)?;
+        self.fd.set_regs(&regs)?;
+
+        dir.set_file_name("kvm_xsave.json");
+        let reader = BufReader::new(File::open(dir.as_path())?);
+        let region_vec: Vec<u32> = serde_json::from_reader(reader)?;
+        let mut region = [0u32; 1024usize];
+        for idx in 0..region.len() {
+            region[idx] = region_vec[idx];
+        }
+        let xsave = kvm_xsave{ region };
+        self.fd.set_xsave(&xsave)?;
+        
+        dir.set_file_name("kvm_xcrs.json");
+        let reader = BufReader::new(File::open(dir.as_path())?);
+        let xcrs: kvm_xcrs = serde_json::from_reader(reader)?;
+        self.fd.set_xcrs(&xcrs)?;
+
+        dir.set_file_name("kvm_sregs.json");
+        let reader = BufReader::new(File::open(dir.as_path())?);
+        let sregs: kvm_sregs = serde_json::from_reader(reader)?;
+        self.fd.set_sregs(&sregs)?;
+
+        self.setup_msrs_from_file(dir)?;
+
+        dir.set_file_name("kvm_vcpu_events.json");
+        let reader = BufReader::new(File::open(dir.as_path())?);
+        let vcpu_events: kvm_vcpu_events = serde_json::from_reader(reader)?;
+        self.fd.set_vcpu_events(&vcpu_events)?;
+
+        dir.set_file_name("kvm_mp_state.json");
+        let reader = BufReader::new(File::open(dir.as_path())?);
+        let mp_state: kvm_mp_state = serde_json::from_reader(reader)?;
+        self.fd.set_mp_state(&mp_state)?;
+
+        dir.set_file_name("kvm_lapic.json");
+        let reader = BufReader::new(File::open(dir.as_path())?);
+        let regs_vec: Vec<std::os::raw::c_char> = serde_json::from_reader(reader)?;
+        let mut regs = [0 as std::os::raw::c_char; 1024usize];
+        for (idx, _) in regs_vec.iter().enumerate() {
+            regs[idx] = regs_vec[idx];
+        }
+        let lapic = kvm_lapic_state { regs };
+        self.fd.set_lapic(&lapic)?;
+
+        dir.pop();
+        Ok(())
+    }
+
+    fn write_regs_to_file(&self, dir: &mut PathBuf) -> result::Result<(), io::Error> {
```

Comment:
rather than use a directory, why not serialize everything into a single file? It could be a single struct with fields for each of the subcomponents you're serializing here, or you could have fixed offsets or something.

But managing a single file is going to be simpler than managing a directory, and overall more expensive, since it'll involve a system call for each data structure.

Reply:
> But managing a single file is going to be simpler than managing a directory, and overall more expensive, since it'll involve a system call for each data structure.

Why managing a single file is more expensive? I thought it was the opposite.

Reply:
I am working towards having one single struct `struct Snapshot` to hold all states but memory and (de)serialize it to (load)dump states. And this will lead to a version that supports multiple vCPUs since on my way I will place capturing of irqchip and memory in Vm instead of in Vcpu because they are per-machine states and originally Vcpu did not hold a vm fd and a reference to memory.

Reply:
Yes, I meant to write _less_ expensive, sorry.